loop_overlay2 driver for stateless image stores

bacaldwell opened the pull request:
Uses loopback devices for layers that are stacked for the
container using overlayfs. This enables support for CoW-like
image layers on a shared file system that lacks overlayfs support.

Driver also backs the image and layer stores using loopback devices
that can reside on shared storage. On graphdriver initialization,
stores found in the root or fallback directories are merged together
using overlayfs, or if only one of the two directories are writable,
then the loopback device is bind mounted. This means changes to the
stores are written and persisted to a loopback device, which is
chosen in order of preference:
  1) loopback file in root directory
  2) loopback file in fallback directory

Additionaly using a loopback device on shared and read-only storage
as the lower dir in an overlay mount provides a consistent view of
already available images across multiple systems. These systems have
"stateless" image stores in the sense that the Docker daemon could
operate with no persistent local storage. The RW working layer of
containers will not be persisted unless an explicit 'docker commit'
is made by the user.

Added a AttachROLoopDevice method to pkg/loopback

Closes #29080 

Signed-off-by: Blake Caldwell <blakec@ornl.gov>

AkihiroSuda commented:
Can it be implemented as a 3rd party graphdriver plugin?

bacaldwell commented:
It's worth considering, but my belief is that it best fits as a core graphdriver. I'd welcome arguments to the contrary, but my reasons are the following:

- It is tightly integrated with docker components beyond the Plugin API. It makes use of Layer and Image Stores for creating the upper (fallback) set of stores, which it then merges with the lower (root) stores.
- It requires new exported methods in pkg/loopback: GetLoopDeviceFor and AttachROLoopDevice. These methods are part of this pull request. Note, this does not preclude a plugin, but is another indicator of the level of integration with Docker's code base.
- It will benefit from unit and integration testing as interfaces such as the Layer and Image Stores evolve.

More conceptual arguments for its inclusion:
- Inclusion of this driver expands Docker use cases in cluster environments: HPC centers (my affiliation is with one), university/scientific clusters, analytics pipelines.
- This driver is vendor and filesystem technology agnostic (although not Windows). It can run on any filesystem (even local /), so applicability is almost as broad as dm-loop, especially over time as overlayfs is supported on more default kernels.
- The feature of a shared image store has been a common request, which argues for the case that it is not a niche use case.

bacaldwell commented:
ping @crosbymichael @thaJeztah

crosbymichael commented:
Interesting.  Let me look more into the design and implementation tomorrow to show how this will be used.

thaJeztah commented:
/cc @dmcgowan

crosbymichael commented:
@bacaldwell in my testing the problem that I have found is the RW layer.  You can store all the RO or `lowerdirs` on things like NFS with no problem, its when you have the `upperdir` on NFS (or others) that overlay won't allow it.

Is this the same with your testing or filesystems that you are using?

dmcgowan commented:
Shared layer storage is very interesting and valuable. My own attempts at playing around with it so far have been done one level up from the graph driver as an alternate implementation of the layer store. I think the separation of the interfaces is important here because we cannot allow the graph drivers to break up a level to know about or alter the layer store. Additionally, graph drivers are more designed for dumb local storage, they have no concept of content addressability making share storage much harder to do with its interface. 

> Can it be implemented as a 3rd party graphdriver plugin?

I think it was a mistake allowing graph driver plugins. It is the wrong abstraction layer and I wouldn't suggest using it. I think we need to develop a higher level image/layer storage pluggable interface which could allow for these more advanced use cases. 

So in summary, :-1: on design as a graph driver, :+1: for further design at a higher level.

bacaldwell commented:
@crosbymichael I understand exactly the issue you are talking about, and it's my oversight for not testing again on shared storage after a significant refactor--I was testing using read-only bind mounts (from root XFS root file system). My intended design was to leave the RW layer as a native directory (not a loopback device).

I will look into this more, but appears that ApplyDiff gets called for the RW layer, and in turn creates a loopback device. Then Get doesn't know which layer is RW.

bacaldwell commented:
@dmcgowan True. Let's discuss the interaction needed by this graph driver.

I would definitely like to see the code to read/merge/write the repositories.json file moved out of the graphdriver. My Driver.collectRepositories method reads repositories.json. Perhaps some exported functions could abstract any dependencies on the JSON-formatting.

The remaining interface to image/layer stores is quite limited:
layer.NewFSMetadataStore
layer.NewStoreFromGraphDriver
image.NewFSStoreBackend
image.NewImageStore

After creation, or identification of existing stores, there is no reading of the stores, just an overlay mount before daemon initialization continues. The use of overlayfs for merging stores depends on the graph driver's configured root_dir and fallback_dir.

Thoughts?

bacaldwell commented:
@crosbymichael I misunderstood the question, sorry. This PR avoids the problem you are talking about. It uses XFS loopback devices on NFS/Lustre/GPFS, which are then stacked using overlayfs. Changes to the RW layer are just file writes inside `upperdir`, which is in `/var/lib/docker/{graphdriver}/`. This directory is **not** shared. If the user wants to make the layer shared, then they would run `docker commit`. Thus RW layer is specific to the Docker daemon, avoiding inconsistencies between different daemons.

My misunderstanding was that I thought you encountered an error with this PR. I've tested with NFS as follows and behavior is as expected.

I'm using dind with `make BIND_DIR=. shell`. I've modified the Makefile to add these bind mounts (unnecessary without dind test environment):
`-v /dev:/dev -v /var/lib/docker/loopback/:/var/lib/docker/loopback/`

On the host `/var/lib/docker/loopback` is a NFS mount, and is where the loopback XFS devices are stored (specifically under a root or fallback directory, whichever one is writable)

bacaldwell committed:
Shared storage graphdriver

Uses loopback devices for layers that are stacked for the
container using overlayfs. This enables support for CoW-like
image layers on a shared file system that lacks overlayfs support.

Driver also backs the image and layer stores using loopback devices
that can reside on shared storage. On graphdriver initialization,
stores found in the root or fallback directories are merged together
using overlayfs, or if only one of the two directories are writable,
then the loopback device is bind mounted. This means changes to the
stores are written and persisted to a loopback device, which is
chosen in order of preference:
  1) loopback file in root directory
  2) loopback file in fallback directory

Additionaly using a loopback device on shared and read-only storage
as the lower dir in an overlay mount provides a consistent view of
already available images across multiple systems. These systems have
"stateless" image stores in the sense that the Docker daemon could
operate with no persistent local storage. The RW working layer of
containers will not be persisted unless an explicit 'docker commit'
is made by the user.

Added some new loopback helpers:
  1) New AttachROLoopDevice method
  2) Added bool parameter "quiet" for getLoopbackBackingFile
  3) New FindLoopDeviceForPath method that calls
     getLoopbackBackingFile with quiet=true

Signed-off-by: Blake Caldwell <blakec@ornl.gov>

bacaldwell commented:
@dmcgowan I made changes that should address your mentioned concerns. Let me know what you think.

Changes:
 - No calls to inside image or layer packages (no longer imported)
 - Doesn't read repositories.json anymore. I omitted the functionality to merge the contents of two json files since it was just a nicety for my use case, not necessary. Overlayfs handles the needed "merged view" of images
 - This graphdriver just prepares access to the storage for the image/layer stores by modifying mountpoints. Towards that end, it needs to know where the daemon expects the image/layer stores to exist (e.g. "/var/lib/docker/image")
   - If needed (not sure the exact scenario for this case--potential to remove), it can be changed with `--storage-opt sharedloop.default_storedir=[val]`

Also changed the driver name to sharedloop to address golint complaints

bacaldwell commented:
Ant feedback on the revised design @dmcgowan @crosbymichael?

LK4D4 commented:
ping @dmcgowan

justincormack commented:
ping @dmcgowan @crosbymichael again - what do you feel about this?

dmcgowan commented:
I stand by my initial assessment that we shouldn't pull this in as a core graphdriver at this time.

@bacaldwell thanks for updating and no longer making calls into layer/image stores. Now that is is more cleanly separated, have you tried to implement it as a plugin? Personally I am not a fan of using the current plugin system for graph drivers, but I think that is the best option for now until we move to go 1.8 and can try dynamic loading for these types of drivers. Additionally, keep your eye on the snapshot work being done at the [containerd repo](https://github.com/docker/containerd/tree/master/snapshot). We intend for that interface to eventually replace graph drivers and want to ensure that interface involves in such a way as enabling these types more advanced use cases.

bacaldwell commented:
I haven't implemented it as a plug-in, but will investigate. Thanks for taking a look at it again.

justincormack commented:
Given the general feeling that we should not take this right now but it could be redone as a plugin, I am going to close this for now. @bacaldwell thanks for your work on this. I do strongly recommend taking a look at the work in `containerd` as I think that this is definitely the sort of thing we should be able to support cleanly there.
